### PR TITLE
KT-20010 'Replace safe access expression with 'if' expression' IDEA Kotlin plugin intention may failed

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-20010

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt [hunk 1]
```diff
@@ -43,12 +47,28 @@ class SafeAccessToIfThenIntention : SelfTargetingRangeIntention<KtSafeQualifiedE
         val dotQualified = psiFactory.createExpressionByPattern("$0.$1", receiver, selector)
 
         val elseClause = if (element.isUsedAsStatement(element.analyze())) null else psiFactory.createExpression("null")
-        val ifExpression = element.convertToIfNotNullExpression(receiver, dotQualified, elseClause)
+        var ifExpression = element.convertToIfNotNullExpression(receiver, dotQualified, elseClause)
+
+        var isEqualExpression = false
+        val binaryExpression = (ifExpression.parent as? KtParenthesizedExpression)?.parent as? KtBinaryExpression
+        if (binaryExpression != null) {
+            val eq = (binaryExpression.operationToken as? KtSingleValueToken)?.takeIf { it == KtTokens.EQ }?.value
```
Comment (yole) [inline idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt]: Why so complicated? `eq` will be always equal to `=`.

File: idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt [hunk 2]
```diff
@@ -43,12 +47,28 @@ class SafeAccessToIfThenIntention : SelfTargetingRangeIntention<KtSafeQualifiedE
         val dotQualified = psiFactory.createExpressionByPattern("$0.$1", receiver, selector)
 
         val elseClause = if (element.isUsedAsStatement(element.analyze())) null else psiFactory.createExpression("null")
-        val ifExpression = element.convertToIfNotNullExpression(receiver, dotQualified, elseClause)
+        var ifExpression = element.convertToIfNotNullExpression(receiver, dotQualified, elseClause)
+
+        var isEqualExpression = false
```
Comment (yole) [inline idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt]: Better call this `isAssignment`

File: idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt [hunk 3]
```diff
@@ -43,12 +47,28 @@ class SafeAccessToIfThenIntention : SelfTargetingRangeIntention<KtSafeQualifiedE
...
+        val binaryExpression = (ifExpression.parent as? KtParenthesizedExpression)?.parent as? KtBinaryExpression
+        if (binaryExpression != null) {
+            val eq = (binaryExpression.operationToken as? KtSingleValueToken)?.takeIf { it == KtTokens.EQ }?.value
+            val right = binaryExpression.right?.text
+            if (eq != null && right != null) {
+                val replaced = binaryExpression.replaced(psiFactory.createExpression("${ifExpression.text} $eq $right"))
```
Comment (yole) [inline idea/src/org/jetbrains/kotlin/idea/intentions/branchedTransformations/intentions/SafeAccessToIfThenIntention.kt]: Better use `createExpressionByPattern`

**Comment by t-kameyama** [comment]
Fixed

**Comment by yole** [comment]
Thanks a lot for your contribution!

**Pull request merged**